File: packages/desktop/lib/features/web_companion/direct_upload/direct_upload_controller.dart
import '../../../core/sidecar/sidecar_api.dart';
import 'direct_upload_models.dart';

/// Thin typed wrapper around [SidecarApi] for the QR upload flow.
///
/// Keeps widgets decoupled from the raw HTTP plumbing so widget tests can
/// inject fakes without touching the real network layer.
class DirectUploadController {
  DirectUploadController({
    required this.api,
    this.serviceUnavailableMessage =
        'QR upload is temporarily unavailable. Please try again later.',
    this.sessionIncompleteMessage =
        'QR upload session could not be created. Check upload settings and try again.',
    this.configIncompleteMessage =
        'QR upload settings are incomplete. Check upload settings and try again.',
  });

  final SidecarApi api;
  final String serviceUnavailableMessage;
  final String sessionIncompleteMessage;
  final String configIncompleteMessage;

  static const _basePath = '/api/web-companion/direct-upload/sessions';

  Future<DirectUploadConfig> createSession(String childId) async {
    final response = await api.post(_basePath, {'childId': childId});
    _throwIfSidecarError(response);
    _requireFields(response, const [
      'sessionId',
      'childId',
      'bucket',
      'sessionPath',
      'supabaseUrl',
      'anonKey',
      'publicUrl',
    ]);
    return DirectUploadConfig.fromJson(response);
  }

  Future<DirectUploadStatusSnapshot> fetchStatus(String sessionId) async {
    final response = await api.get('$_basePath/$sessionId/status');
    return DirectUploadStatusSnapshot.fromJson(response);
  }

  Future<DirectUploadStatusSnapshot> triggerPullback(
    String sessionId, {
    required String token,
    List<String>? objectKeys,
  }) async {
    final body = <String, dynamic>{'token': token};
    if (objectKeys != null) {
      body['objectKeys'] = objectKeys;
    }
    final response = await api.post('$_basePath/$sessionId/pullback', body);
    _throwIfSidecarError(response);
    return fetchStatus(sessionId);
  }

  void _throwIfSidecarError(Map<String, dynamic> response) {
    if (response['ok'] == false) {
      final message = response['message'];
      final code = response['code'];
      if (message is String && message.trim().isNotEmpty) {
        throw StateError(_ordinaryErrorMessage(message));
      }
      if (code is String && code.trim().isNotEmpty) {
        throw StateError(_ordinaryErrorMessage(code));
      }
      throw StateError(serviceUnavailableMessage);
    }
  }

  void _requireFields(Map<String, dynamic> response, List<String> keys) {
    final missing = keys
        .where((key) {
          final value = response[key];
          return value == null || (value is String && value.trim().isEmpty);
        })
        .toList(growable: false);
    if (missing.isNotEmpty) {
      throw StateError(sessionIncompleteMessage);
    }
  }

  String _ordinaryErrorMessage(String message) {
    final lower = message.toLowerCase();
    if (lower.contains('supabase') ||
        lower.contains('sidecar') ||
        lower.contains('requestid') ||
        lower.contains('jobid') ||
        lower.contains('anon_key') ||
        lower.contains('service_role')) {
      return configIncompleteMessage;
    }
    return message;
  }
}

File: packages/backend/src/modules/web-companion/direct-upload.controller.ts
/**
 * Web Companion Supabase Direct Upload controller.
 *
 * 路由（与 trusted upload 隔离，挂在 `/web-companion/direct-upload/*` 命名空间）：
 *   POST   /web-companion/direct-upload/sessions
 *   GET    /web-companion/direct-upload/sessions/:sessionId/objects
 *   POST   /web-companion/direct-upload/sessions/:sessionId/pullback
 *   GET    /web-companion/direct-upload/sessions/:sessionId/status
 *   GET    /web-companion/direct-upload/sessions/:sessionId/config
 */

import { Body, Controller, Get, Inject, Param, Post } from "@nestjs/common";

import { DirectUploadService } from "./direct-upload.service.ts";
import type {
  CreateDirectUploadSessionRequest,
  CreateDirectUploadSessionResponse,
} from "./dto/create-direct-upload-session.dto.ts";
import type { ListDirectUploadObjectsResponse } from "./dto/list-direct-upload-objects.dto.ts";
import type {
  PullbackDirectUploadRequest,
  PullbackDirectUploadResponse,
} from "./dto/pullback-direct-upload.dto.ts";
import type { GetDirectUploadStatusResponse } from "./dto/get-direct-upload-status.dto.ts";

export class DirectUploadController {
  private readonly service: DirectUploadService;

  constructor(service: DirectUploadService) {
    this.service = service;
  }

  createSession(
    body: CreateDirectUploadSessionRequest,
  ): Promise<CreateDirectUploadSessionResponse> {
    return this.service.createSession(body || { childId: "" });
  }

  listObjects(sessionId: string): Promise<ListDirectUploadObjectsResponse> {
    return this.service.listObjects(sessionId);
  }

  pullback(
    sessionId: string,
    body: PullbackDirectUploadRequest,
  ): Promise<PullbackDirectUploadResponse> {
    return this.service.pullback(sessionId, body || {});
  }

  getStatus(sessionId: string): Promise<GetDirectUploadStatusResponse> {
    return this.service.getStatus(sessionId);
  }

  getSessionConfig(
    sessionId: string,
  ): Promise<{ supabaseUrl: string; anonKey: string; bucket: string; recommendedClientLimit: number }> {
    return this.service.getSessionConfig(sessionId);
  }
}

// ---- Nest 装饰器手动注册 (与 web-companion.controller.ts 保持一致风格) ------

Inject(DirectUploadService)(DirectUploadController, undefined, 0);
Controller("api/web-companion/direct-upload")(DirectUploadController);

Post("sessions")(
  DirectUploadController.prototype,
  "createSession",
  Object.getOwnPropertyDescriptor(DirectUploadController.prototype, "createSession")!,
);
Body()(DirectUploadController.prototype, "createSession", 0);

Get("sessions/:sessionId/objects")(
  DirectUploadController.prototype,
  "listObjects",
  Object.getOwnPropertyDescriptor(DirectUploadController.prototype, "listObjects")!,
);
Param("sessionId")(DirectUploadController.prototype, "listObjects", 0);

Post("sessions/:sessionId/pullback")(
  DirectUploadController.prototype,
  "pullback",
  Object.getOwnPropertyDescriptor(DirectUploadController.prototype, "pullback")!,
);
Param("sessionId")(DirectUploadController.prototype, "pullback", 0);
Body()(DirectUploadController.prototype, "pullback", 1);

Get("sessions/:sessionId/status")(
  DirectUploadController.prototype,
  "getStatus",
  Object.getOwnPropertyDescriptor(DirectUploadController.prototype, "getStatus")!,
);
Param("sessionId")(DirectUploadController.prototype, "getStatus", 0);

Get("sessions/:sessionId/config")(
  DirectUploadController.prototype,
  "getSessionConfig",
  Object.getOwnPropertyDescriptor(DirectUploadController.prototype, "getSessionConfig")!,
);
Param("sessionId")(DirectUploadController.prototype, "getSessionConfig", 0);
